File: android/app/src/main/java/com/sanai/LlamaEngineModule.java
package com.sanai;

import android.util.Log;
import androidx.annotation.NonNull;

import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.HashMap;
import java.util.Map;

public class LlamaEngineModule extends ReactContextBaseJavaModule {
    private static final String TAG = "LlamaEngineModule";
    private static final String EVENT_TOKEN = "onToken";
    private static final String EVENT_COMPLETE = "onComplete";
    private static final String EVENT_ERROR = "onError";
    private static final String EVENT_PROGRESS = "onProgress";
    private static final String EVENT_TOOL_CALL = "onToolCall";

    private long context = 0;
    private boolean isGenerating = false;

    public LlamaEngineModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @NonNull
    @Override
    public String getName() {
        return "LlamaEngineModule";
    }

    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        return constants;
    }

    @ReactMethod
    public void initialize(ReadableMap config, Promise promise) {
        try {
            String modelPath = config.getString("modelPath");
            int contextLength = config.getInt("contextLength");
            int threads = config.getInt("threads");
            int gpuLayers = config.getInt("gpuLayers");

            Log.d(TAG, "Initializing with model: " + modelPath);
            Log.d(TAG, "Context length: " + contextLength + ", Threads: " + threads);

            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Initialization failed", e);
            promise.reject("INIT_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void loadModel(String modelPath, ReadableMap config, Promise promise) {
        try {
            int contextLength = config.hasKey("contextLength") ? config.getInt("contextLength") : 4096;
            int threads = config.hasKey("threads") ? config.getInt("threads") : 4;
            int gpuLayers = config.hasKey("gpuLayers") ? config.getInt("gpuLayers") : 35;

            Log.d(TAG, "Loading model: " + modelPath);
            promise.resolve(true);
        } catch (Exception e) {
            Log.e(TAG, "Model loading failed", e);
            promise.reject("LOAD_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void unloadModel(Promise promise) {
        try {
            context = 0;
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("UNLOAD_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void generate(String prompt, ReadableMap config, Promise promise) {
        try {
            double temperature = config.hasKey("temperature") ? config.getDouble("temperature") : 0.7;
            double topP = config.hasKey("topP") ? config.getDouble("topP") : 0.9;
            int topK = config.hasKey("topK") ? config.getInt("topK") : 40;
            double repeatPenalty = config.hasKey("repeatPenalty") ? config.getDouble("repeatPenalty") : 1.1;
            int maxTokens = config.hasKey("maxTokens") ? config.getInt("maxTokens") : 2048;

            isGenerating = true;

            Log.d(TAG, "Generating with temperature: " + temperature);
            promise.resolve(true);
        } catch (Exception e) {
            isGenerating = false;
            promise.reject("GENERATE_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void stopGeneration(Promise promise) {
        try {
            isGenerating = false;
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("STOP_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void getRamUsage(Promise promise) {
        try {
            Runtime runtime = Runtime.getRuntime();
            long usedMemory = (runtime.totalMemory() - runtime.freeMemory()) / (1024 * 1024);
            long maxMemory = runtime.maxMemory() / (1024 * 1024);

            WritableMap result = Arguments.createMap();
            result.putInt("used", (int) usedMemory);
            result.putInt("total", (int) maxMemory);
            promise.resolve(result);
        } catch (Exception e) {
            promise.reject("RAM_ERROR", e.getMessage());
        }
    }

    private void sendEvent(String eventName, String data) {
        try {
            ReactContext reactContext = getReactApplicationContext();
            WritableMap params = Arguments.createMap();
            params.putString("type", eventName);
            params.putString("data", data);
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
        } catch (Exception e) {
            Log.e(TAG, "Failed to send event: " + eventName, e);
        }
    }

    private void sendEvent(String eventName, WritableMap data) {
        try {
            ReactContext reactContext = getReactApplicationContext();
            WritableMap params = Arguments.createMap();
            params.putString("type", eventName);
            params.putMap("data", data);
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
        } catch (Exception e) {
            Log.e(TAG, "Failed to send event: " + eventName, e);
        }
    }
}

File: ios/sanAi/LlamaEngineModule.m
import Foundation
import React

@objc(LlamaEngineModule)
class LlamaEngineModule: RCTEventEmitter {

    private var context: OpaquePointer?
    private var model: OpaquePointer?
    private var isGenerating = false

    override static func requiresMainQueueSetup() -> Bool {
        return false
    }

    override func supportedEvents() -> [String]! {
        return ["onToken", "onComplete", "onError", "onProgress", "onToolCall"]
    }

    @objc(initialize:withResolver:withRejecter:)
    func initialize(_ config: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        guard let modelPath = config["modelPath"] as? String else {
            reject("INIT_ERROR", "Model path is required", nil)
            return
        }

        let contextLength = config["contextLength"] as? Int32 ?? 4096
        let threads = config["threads"] as? Int32 ?? 4
        let gpuLayers = config["gpuLayers"] as? Int32 ?? 35

        // TODO: Initialize llama.cpp
        // var model_params = llama_model_default_params()
        // model_params.n_gpu_layers = gpuLayers
        // model = llama_load_model_from_file(modelPath, model_params)

        print("LlamaEngine: Initializing model at \(modelPath)")
        print("LlamaEngine: Context length: \(contextLength), Threads: \(threads), GPU layers: \(gpuLayers)")

        resolve(true)
    }

    @objc(loadModel:config:withResolver:withRejecter:)
    func loadModel(_ modelPath: String, config: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let contextLength = config["contextLength"] as? Int32 ?? 4096
        let threads = config["threads"] as? Int32 ?? 4

        // TODO: Load model using llama.cpp
        // var ctx_params = llama_context_default_params()
        // ctx_params.n_ctx = contextLength
        // ctx_params.n_threads = threads

        print("LlamaEngine: Loading model \(modelPath)")
        resolve(true)
    }

    @objc(unloadModel:withRejecter:)
    func unloadModel(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        // TODO: Free llama context and model
        // llama_free(context)
        // llama_free_model(model)
        context = nil
        model = nil
        resolve(nil)
    }

    @objc(generate:config:withResolver:withRejecter:)
    func generate(_ prompt: String, config: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let temperature = config["temperature"] as? Float ?? 0.7
        let topP = config["topP"] as? Float ?? 0.9
        let topK = config["topK"] as? Int32 ?? 40
        let maxTokens = config["maxTokens"] as? Int32 ?? 2048

        isGenerating = true

        // TODO: Run inference in background
        // DispatchQueue.global(qos: .userInitiated).async {
        //     while self.isGenerating {
        //         // Generate tokens and send via sendEvent
        //     }
        // }

        print("LlamaEngine: Generating with temperature: \(temperature)")
        resolve(true)
    }

    @objc(stopGeneration:withRejecter:)
    func stopGeneration(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        isGenerating = false
        resolve(nil)
    }

    @objc(getRamUsage:withRejecter:)
    func getRamUsage(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let processInfo = ProcessInfo.processInfo
        let physicalMemory = processInfo.physicalMemory / (1024 * 1024)

        resolve([
            "used": 0,
            "total": Int(physicalMemory)
        ])
    }

    private func sendEvent(name: String, body: Any?) {
        self.sendEvent(withName: name, body: body)
    }
}
